### PR TITLE
Feature/auth admin api impl

### PR DESCRIPTION
**Pull request by unnarb**
#Merging Auth related services and library

Attach a link to issue if relevant

## What

Merging auth libs and services to master

## Why

We need to merge these changes to master before other major changes occure in Master branch

## Screenshots / Gifs

Attach Screenshots / Gifs to help reviewers understand the scope of the pull request

## Checklist:

- [x ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ X] Formatting passes locally with my changes
- [ ] I have rebased against master before asking for a review

### CODE REVIEW

**Review by valurefugl — APPROVED** [review]

File: workspace.json [hunk 1]
```diff
@@ -3066,6 +3094,172 @@
...
+              "**/node_modules/**",
+              "!apps/services/auth-admin-api/**"
+            ]
+          }
+        },
+        "test": {
+          "builder": "@nrwl/jest:jest",
+          "options": {
+            "jestConfig": "apps/services/auth-admin-api/jest.config.js",
+            "tsConfig": "apps/services/auth-admin-api/tsconfig.spec.json",
```
Comment (eirikurn) [inline workspace.json:3151]: Ditto

File: workspace.json [hunk 2]
```diff
@@ -3066,6 +3094,172 @@
...
+              "apps/services/auth-api/tsconfig.spec.json"
+            ],
+            "exclude": ["**/node_modules/**", "!apps/services/auth-api/**"]
+          }
+        },
+        "test": {
+          "builder": "@nrwl/jest:jest",
+          "options": {
+            "jestConfig": "apps/services/auth-api/jest.config.js",
+            "tsConfig": "apps/services/auth-api/tsconfig.spec.json",
```
Comment (eirikurn) [inline workspace.json:3208]: Ditto

File: workspace.json [hunk 3]
```diff
@@ -3066,6 +3094,172 @@
...
+          "options": {
+            "commands": [
+              {
+                "command": "docker-compose -f docker-compose.base.yml -f docker-compose.dev.yml up -d"
+              }
+            ],
+            "cwd": "apps/services/auth-api"
+          }
+        },
+        "generate-schema": {
```
Comment (eirikurn) [inline workspace.json:3224]: We might want to rename this to `init-schema` so we can gitignore the schema and generate it in `postinstall` and in `CI`.

I see that this is currently pretty random and confusing, looking across workspace.json, but we plan to simplify it a bunch (see 
#1225).

**Review by eirikurn — DISMISSED** [review]
Wow. That was a long read :)

But awesome job! Overall, this looks really good. Very clean code and great use of nest. Love the shared lib setup.

I only scanned the code and did not dig deep into the architecture or security concerns. I would love to see a database diagram somewhere, e.g. in api-auth-lib readme or Notion.

Look forward to getting this merged!

File: apps/services/auth-admin-api/src/app/modules/resources/resources.controller.ts [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+  // @ApiOkResponse({ type: IdentityResource })
+  // async updateIdentityResource(
+  //   @Body() identityResource: IdentityResourcesDTO,
+  //   @Param('id') id: string,
+  // ): Promise<IdentityResource> {
+  //   return await this.resourcesService.updateIdentityResource(
+  //     identityResource,
+  //     id,
+  //   )
+  // }
```
[the remark this answers is not in the document]
Comment (eirikurn) [reply apps/services/auth-admin-api/src/app/modules/resources/resources.controller.ts:95]: (Feel free to resolve these if you mean to address this in a follow-up.)

File: libs/auth-api-lib/migrations/20200814053200-create-user-identity.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
+'use strict'
+/* eslint-env node */
+
+module.exports = {
+  up: (queryInterface) => {
+    return queryInterface.sequelize.query(`
```
[the remark this answers is not in the document]
Comment (unnarb) [reply libs/auth-api-lib/migrations/20200814053200-create-user-identity.js:6]: Is it a problem to keep these migrations in the library?

File: apps/services/auth-api/docker-compose.base.yml [hunk 6]
```diff
@@ -0,0 +1,22 @@
+version: '3.3'
+
+services:
+  db:
+    image: postgres:11.6
+    container_name: db
+    networks:
+      - local
+    environment:
+      - POSTGRES_PASSWORD=XjFsvIrGyCCrdE
```
[the remark this answers is not in the document]
Comment (unnarb) [reply apps/services/auth-api/docker-compose.base.yml:10]: Fixed

File: libs/auth-api-lib/src/lib/auth-api.spec.ts [hunk 7]
```diff
@@ -0,0 +1,7 @@
+import { authApi } from './auth-api'
+
+describe('authApi', () => {
+  it('should work', () => {
+    expect(authApi()).toEqual('auth-api')
+  })
+})
```
[the remark this answers is not in the document]
Comment (unnarb) [reply libs/auth-api-lib/src/lib/auth-api.spec.ts:7]: Fixed

File: libs/auth-api-lib/src/lib/auth-api.ts [hunk 8]
```diff
@@ -0,0 +1,3 @@
+export function authApi(): string {
+  return 'auth-api'
+}
```
[the remark this answers is not in the document]
Comment (unnarb) [reply libs/auth-api-lib/src/lib/auth-api.ts:3]: fixed

File: libs/infra-nest-server/src/lib/bootstrap.ts [hunk 9]
```diff
@@ -17,6 +17,13 @@ import yaml from 'js-yaml'
 import * as yargs from 'yargs'
 import * as fs from 'fs'
 
+// const AuthUrl = "https://siidentityserverweb20200805020732.azurewebsites.net/connect/authorize"
+// const TokenUrl = "https://siidentityserverweb20200805020732.azurewebsites.net/connect/token"
+// const AuthUrl = `${process.env.IDS_ISSUER}/connect/authorize`
+// const TokenUrl = `${process.env.IDS_ISSUER}/connect/token`
+const AuthUrl = `https://localhost:6001/connect/authorize`
+const TokenUrl = `https://localhost:6001/connect/token`
```
[the remark this answers is not in the document]
Comment (unnarb) [reply libs/infra-nest-server/src/lib/bootstrap.ts:25]: Fyrir swagger. @stullifugl geturðu skoðað

File: libs/infra-nest-server/src/lib/bootstrap.ts [hunk 10]
```diff
@@ -7,7 +7,7 @@ import {
   ValidationPipe,
   NestInterceptor,
 } from '@nestjs/common'
-import { OpenAPIObject, SwaggerModule } from '@nestjs/swagger'
+import { OpenAPIObject, SwaggerModule, DocumentBuilder } from '@nestjs/swagger'
```
[the remark this answers is not in the document]
Comment (unnarb) [reply libs/infra-nest-server/src/lib/bootstrap.ts:10]: Fyrir swagger @stullifugl ?

File: libs/auth-api-lib/src/lib/services/user-identities.service.ts [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+import { UserIdentity } from '../entities/models/user-identity.model'
+import { UserIdentityDto } from '../entities/dto/user-identity.dto'
+
+@Injectable()
+export class UserIdentitiesService {
+  applicationsRegistered = new Counter({
+    name: 'apps_registered2',
+    labelNames: ['res1'],
+    help: 'Number of applications',
+  }) // TODO: How does this work?
```
[the remark this answers is not in the document]
Comment (pshomov) [reply libs/auth-api-lib/src/lib/services/user-identities.service.ts:17]: Sorry got the notification about this but was in the middle of a release. 
`labelNames` are not required, so they can be skipped altogether. For more information about metrics see here - https://github.com/island-is/handbook/blob/master/metrics.md

[the remark this answers is not in the document]
Comment (pshomov) [reply libs/auth-api-lib/src/lib/services/user-identities.service.ts:17]: If something is unclear about metrics after reading the docs, would love to get comments to improve it.

File: apps/services/auth-admin-api/src/app/modules/resources/resources.controller.ts [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+  IdentityResource,
+  ResourcesService,
+  ApiScope,
+  ApiResource,
+  ApiScopesDTO,
+  IdentityResourcesDTO,
+} from '@island.is/auth-api-lib'
+import { AuthGuard } from '@nestjs/passport'
+
+@ApiOAuth2(['@identityserver.api/read'])
```
[the remark this answers is not in the document]
Comment (valurefugl) [reply apps/services/auth-admin-api/src/app/modules/resources/resources.controller.ts:29]: Added guard that checks scope.

File: apps/services/auth-api/docker-compose.base.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (unnarb) [reply apps/services/auth-api/docker-compose.base.yml:10]: Comes from other projects that we based on

File: libs/auth-api-lib/migrations/20200814053200-create-user-identity.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (unnarb) [reply libs/auth-api-lib/migrations/20200814053200-create-user-identity.js:6]: We will look into how is best to migrate on deployment

**Review by arnifreyrs — DISMISSED** [review]
Great work guys! I tried to provide some feedback, I feel like all of Eiki's feedback has been addressed. 

My only concerns is that for the average developer, it will be very beneficial to have better README content to understand this better. At least the modules that will be part of everyday development in this repo.

Also, there are no tests whatsoever. Is that something that will be addressed soon? I know we need this PR in the repo asap, so adding tests could be part of the next PRs.

**Review by valurefugl — APPROVED** [review]

**Review by arnifreyrs — APPROVED** [review]

**Review by arnifreyrs — APPROVED** [review]

**Review by arnifreyrs — APPROVED** [review]

**Pull request merged**